staging/src/k8s.io/apiserver/pkg/admission/plugin/cel/admission.go

```diff
@@ -53,29 +58,72 @@ func Register(plugins *admission.Plugins) {
 ////////////////////////////////////////////////////////////////////////////////
 
 type celAdmissionPlugin struct {
-	evaluator CELPolicyEvaluator
+	evaluator             CELPolicyEvaluator
+	inspectedFeatureGates bool
+	enabled               bool
+	client                kubernetes.Interface
+	namespaceInformer     informers.SharedInformerFactory
 }
 
+var _ initializer.WantsExternalKubeInformerFactory = &celAdmissionPlugin{}
+var _ initializer.WantsExternalKubeClientSet = &celAdmissionPlugin{}
 var _ WantsCELPolicyEvaluator = &celAdmissionPlugin{}
+var _ admission.InitializationValidator = &celAdmissionPlugin{}
 var _ admission.ValidationInterface = &celAdmissionPlugin{}
 
-func NewPlugin() (*celAdmissionPlugin, error) {
+func NewPlugin() (admission.Interface, error) {
 	result := &celAdmissionPlugin{}
 	return result, nil
 }
 
 func (c *celAdmissionPlugin) SetCELPolicyEvaluator(evaluator CELPolicyEvaluator) {
 	c.evaluator = evaluator
+	if c.client != nil {
+		c.evaluator.SetExternalKubeClientSet(c.client)
+	}
+	if c.namespaceInformer != nil {
+		c.evaluator.SetExternalKubeInformerFactory(c.namespaceInformer)
```

Comment:
Setting this from the plugin means that each instance of the admission plugin might change the factory used by the evaluator to their own copy (which is the same in our case, but generally seems incorrect).

I think the factory and client should be given to the evaluator instead via the `NewAdmissionController` call in `config.go`

Reply:
You hit the question which struggle me the most^^. I guess using `SetExternalKubeClientSet` and `SetExternalKubeInformerFactory` is the recommended way. But evaluator could be nil at this point and we cannot guarantee the order of being called. All we need is to pass those all the way to matcher and verify while initialization. Since we will not modify factory so I don't have strong preference on either way

Reply:
Btw mostly likely the factory will be passed though `SetExternalKubeInformerFactory`. Here just to add checking in case it is called before SetCELPolicyEvaluator